src/benchmarks/micro/runtime/Interop/ComWrappers.cs

```diff
@@ -7,63 +7,60 @@
-using System.Threading;
 using System.Threading.Tasks;
 
 namespace Interop
 {
-    [BenchmarkCategory(Categories.Runtime)]
-    public unsafe class ComWrappersTests
+    [BenchmarkCategory(Categories.Runtime, Categories.NoWASM)]
+    public class ComWrappersTests
     {
         static readonly ComWrappers s_instance = new DummyComWrappers();
 
-        static object QueryRCW(nint ptr)
-            => s_instance.GetOrCreateObjectForComInstance(ptr, CreateObjectFlags.None);
-
-        private List<nint> iterations;
+        private nint ptr;
         private object targetRcw;
 
         public ComWrappersTests()
         {
-            // Allocate a simple IUnknown instance.
-            var ptr = (void**)NativeMemory.Alloc(5, (nuint)sizeof(void*));
-
-            ptr[1] = (void*)1; // Initial ref count.
-
-            // Define IUnknown vtable.
-            ptr[2] = (delegate* unmanaged<void*, Guid*, void**, int>)&QueryInterface;
-            ptr[3] = (delegate* unmanaged<void*, int>)&AddRef;
-            ptr[4] = (delegate* unmanaged<void*, int>)&Release;
-
-            // Assign vtable to instance.
-            ptr[0] = &ptr[2];
-
-            // Record RCW pointer.
-            nint rcwPtr = (nint)ptr;
+            // Let the ComWrappers API create a valid IUnknown instance.
+            // This can then be passed as an unmanaged instance since ComWrappers
+            // doesn't unwrap by default.
+            object tmp = new object();
+            ptr = s_instance.GetOrCreateComInterfaceForObject(tmp, CreateComInterfaceFlags.None);
 
             // Transfer ownership to RCW instance. This populates the RCW cache.
-            targetRcw = s_instance.GetOrCreateObjectForComInstance((nint)ptr, CreateObjectFlags.None);
+            targetRcw = s_instance.GetOrCreateObjectForComInstance(ptr, CreateObjectFlags.None);
+            if (tmp == targetRcw)
+            {
+                throw new Exception("Test invariant violated. Assuming roundtrip for CCW to RCW doesn't unwrap.");
+            }
 
             // Ownership was transferred.
-            Marshal.Release((nint)ptr);
-
-            // Define a large number of iterations for parallel action.
-            var limit = 3_000_000;
-            iterations = Enumerable.Repeat(rcwPtr, limit).ToList();
+            Marshal.Release(ptr);
         }
 
         [Benchmark]
-        public unsafe ParallelLoopResult ParallelRCWLookUp()
-            => Parallel.ForEach(iterations, static (nint ptr) => QueryRCW(ptr));
+        public async Task ParallelRCWLookUp()
+        {
+            // Define a large number of iterations for parallel action.
+            var count = 3_000_000;
+            await Task.WhenAll(
+                Enumerable.Repeat(ptr, count)
+                    .Select(ptr =>
+                        Task.Run(delegate
+                        {
+                            s_instance.GetOrCreateObjectForComInstance(ptr, CreateObjectFlags.None);
+                        })));
```

Comment:
```suggestion
            await Task.WhenAll(
                Enumerable
                .Range(0, Environment.ProcessorCount)
                .Select(_ =>
                Task.Run(async delegate
                {
                    // Define a large number of iterations for parallel action.
                    for (int i = 0; i < 3_000_000; i++)
                    {
                        s_instance.GetOrCreateObjectForComInstance(ptr, CreateObjectFlags.None);
                    }
                })));
```
Scheduling a task is more expensive operation than `s_instance.GetOrCreateObjectForComInstance`. (On my notebook, the perf tests spends 1.3 seconds in the task scheduler and 0.5 seconds in  `s_instance.GetOrCreateObjectForComInstance` on .NET 8 P7 - before the perf fix.)

The perf test needs to schedule a lot fewer tasks than `s_instance.GetOrCreateObjectForComInstance` calls. Otherwise, it would be testing the perf of the task scheduler and not the perf of the ComWrappers.

Reply:
My suggestion is copy&paste from https://github.com/dotnet/performance/blob/c30478de1d1e8903221d8b21920a9c980d65ee55/src/benchmarks/micro/libraries/System.Buffers/ArrayPoolTests.cs#L66C1-L81C1